[RFC] Precompiled headers for UNIX

neo1973 opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Add support for precompiled headers for UNIX platforms using CMake facilities. The same approach should also work for WIN32 but because WIN32 already has working PCH I want to focus on UNIX for now.

How does this effect build times for a debug build? In my configuration on Linux a build from scratch went from 8:52 min to 8:28 min for GCC and from 7:30 min to 4:24 min for Clang. The Ninja build system was used in all cases. As these were builds from scratch there are many build stages that don't take advantage of PCH (e.g. building external dependencies) so the relative gains are higher when developing with incremental rebuilds.

For now this is only a RFC and I would like to get some feedback if we want PCH at all for UNIX platforms and if yes if it's the right approach. It also would be nice to know if this works on Mac with Xcode.

Some comments on the commits:

- The spdlog commit is only a place holder until #20841 is merged because the header only spdlog isn't PCH friendly due to template bloat.
- The CacheStrategie commit is there because fmt doesn't like the `volatile` when `fmt/std.h` is in the precompiled header. Looks like a fmt bug but because it's unnecessary anyway I removed it.
- The other commits align the compiler options between the different modules. This is necessary to share the PCH between the modules. Otherwise each module would require it's own PCH which creates so much overhead that PCHs become basically useless.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Speeding up compilation to make development more pleasant and speed up Jenkins builds.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Building with and without PCH with Clang and GCC on Linux. I haven't tested a build with all configuration options enabled yet so it's very likely that not all configurations are working as of now.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
None.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

neo1973 committed:
[CMake] Enable SSE/SSE2 globaly if available

neo1973 committed:
[CacheStrategy] Drop volatile key words

This was probably at some point added due to multi threading concerns but
volatile doesn't work that way.

neo1973 committed:
[CMake] Don't enable position indipendent code when we don't build a shared library

neo1973 committed:
[CMake] Add precompiled header support for UNIX platforms

neo1973 committed:
[CMake][libUPnP] Don't leak the NPT_CONFIG_ENABLE_LOGGING flag

It's only required internally when building the library, consumers don't need
it.

wsnipex commented:
precompiled headers is a nice improvement, thanks for working on this!

neo1973 committed:
[EventStream] Change namespace 'detail' to 'EventStreamDetail' to make it unique

neo1973 commented:
One remaining issues is ccache. To play nice with PCH the sloppiness must be set accordingly (https://ccache.dev/manual/4.6.3.html#_precompiled_headers). This should be possible to add to FindCCache.cmake.

garbear commented:
Agree with @wsnipex, great improvement! Mostly I like those enticing numbers showing the speedup.

neo1973 commented:
I just realised that I have ccache installed and it messed with my initial measurements. I updated the PR with the correct numbers. But I really wonder why GCC only sees such a small improvement...

wsnipex commented:
https://github.com/xbmc/xbmc/pull/20841 has been merged

neo1973 committed:
[CMake] Make pre-compiled header play nice with Ccache

neo1973 committed:
[CMake] Don't disable optimizations for Mime.cpp on Clang

Changing the optimization isn't compatible with PCH. And the workaround
probably isn't required anymore anyway.

fuzzard commented:
Not sure if you still wanted to press ahead with this @neo1973 

Ive looked at the apple side of things. Ive got a few commits that makes things work (both make and xcode projects)

This one just disables PCH usage for objective-c/c++ targets/files. Supposedly they should be supported, but its all sorts of fucked. just easier to disable for the handful of targets. They are a very small minority of the build time anyway.
https://github.com/fuzzard/xbmc/commit/bcb2b78bbaef593703a7e0cd4b9171f8d6f56a5a

Xcode specifically doesnt like both the C and C++ pch headers in one file. i poked around at a few ways around this, but in the end i think its just easier to not use a PCH for the C language. Its only required for the CC_Decoder stuff, so again, minimal.
https://github.com/fuzzard/xbmc/commit/c3d555e4f64b5b19ee90bf1288fa10d9e9f60524

libupnp also caused issues. Just easier to not use a PCH there for apple at this point. It does work with make, but xcode has a bit of a hissy fit.
https://github.com/fuzzard/xbmc/commit/d9a7d2aa6eae05899d00eed05d236df6a6600740

The last one is a bit of an odd one. I didnt spend much time on this one, but xcode would get regular failures with timestamp issues when using ccache. Easiest resolution was to add a ``-Xclang -fno-validate-pch`` to the ccache launcher. I'll be honest and say i really am not sure the ramifications, but things build consistently with it, so /shrug
```
Showing Recent Errors Only
File 'build/CMakeFiles/libkodi.dir/Debug/cmake_pch.hxx' has been modified since the precompiled header 'build/build/SharedPrecompiledHeaders/SharedPrecompiledHeaders/14341157800669414611/cmake_pch.hxx.gch' was built: mtime changed (was 1716598331, now 1716602488)
```
https://github.com/fuzzard/xbmc/commit/617b793ec89c3d9877a2b33973b26ba0a92a8361

Full rebased branch at https://github.com/fuzzard/xbmc/commits/precompiled_headers
